guppylang/checker/expr_checker.py

```diff
@@ -295,16 +295,12 @@ def visit_Call(self, node: ast.Call, ty: Type) -> tuple[ast.expr, Subst]:
                     check_inst(f_ty, inst, node)
                     # Expect that each function is a `CallableDef`
                     if isinstance(f_processed, GlobalName):
-                        assert isinstance(
-                            self.ctx.globals[f_processed.def_id], CallableDef
-                        )
-                        call_nodes.append(
-                            GlobalCall(
-                                def_id=f_processed.def_id,
-                                args=processed_args,
-                                type_args=inst,
-                            )
+                        defn = self.ctx.globals[f_processed.def_id]
+                        assert isinstance(defn, CallableDef)
+                        call_node, subst = defn.check_call(
```

Comment:
You're overriding the `subst` you got from `check_call_with_leftovers`. The substitution you get here should always be empty since `f_ty.output` shouldn't have any unsolved vars.